### PR TITLE
fix: Self deletion logic delegated to kalium Use Cases

### PR DESCRIPTION
**Pull request by gongracr**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues
Currently the app layer when sending a message, specifies the expiration data value that will be used to determine when a message should be deleted. Therefore Kalium, just takes this values and assumes it's the correct one, allowing for misusages or human errors when deciding the value to package the message on the protobuf envelope.

### Solutions

Add an extra db query that will update and select the `Message` with the latest self deletion data.

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Review by trOnk12 — APPROVED** [review]

**Review by vitorhugods — COMMENTED** [review]
Why delegate this to the Repository?

Also, I know we don't support this at the moment, but what about offline sending or scheduling of messages?

_i.e._ User tries to send while offline with timer `X`, then changes the timer to `Y`, and then tries to send another message.

When retrying, the timer of both messages will be `Y`, violating the user's will because the timer is figured out on-the-fly.

I would expect the UseCase to contain this logic and whatever the UseCase persists should be the law, and not to be altered by the sending mechanism.

I believe this kind of coupling and doubled responsibility (_i.e._ the interceptor being _too_ smart) could hunt our dreams in the future when we re-enable sending retrying.

**Comment by MohamadJaara** [comment]
@vitorhugods 
The reason for delegating the logic to kalium, so it is global and will be applied even when sending from notifications.

We can move the logic up to the use case, but then the question is if a conversation has a timer of 1m and a user schedules a message for  then the team changes the timer to 4 weeks, should the scheduled message then be 1m or 4w ?

**Comment by vitorhugods** [comment]
> @vitorhugods The reason for delegating the logic to kalium, so it is global and will be applied even when sending from notifications.
> 
> We can move the logic up to the use case, but then the question is if a conversation has a timer of 1m and a user schedules a message for then the team changes the timer to 4 weeks, should the scheduled message then be 1m or 4w ?

Considering the user was offline when it happened, it's understandable something like this could happen, and the design is probably debatable.
In any case, it would be easier to handle such case by updating the pending messages to follow the new team settings once a update event happens, as sending will only happen _after_ all events are processed.

Having it on-the-fly doesn't allow for such a gracious handling of such cases.

**Comment by gongracr** [comment]
> Why delegate this to the Repository?
> 
> Also, I know we don't support this at the moment, but what about offline sending or scheduling of messages?
> 
> _i.e._ User tries to send while offline with timer `X`, then changes the timer to `Y`, and then tries to send another message.
> 
What's the problem there :) ? The message will always use the most recent timer set on the conversation or user settings. Otherwise, as Mo pointed out, we might be hacking our own code/enforced settings.


> When retrying, the timer of both messages will be `Y`, violating the user's will because the timer is figured out on-the-fly.
> 
I don't think this should be violating user's will, the message will be just following the current enforced settings with a little magic under the hood which is not evil per se imho. 


> I would expect the UseCase to contain this logic and whatever the UseCase persists should be the law, and not to be altered by the sending mechanism.
> 
I'm not necessarily against this approach, but following this will imply that we need to replicate this messageDao query for each use case we have now and any future ones that we add in the future (i.e. we decide at some point that reactions or pings should also be self deleted)


> I believe this kind of coupling and doubled responsibility (_i.e._ the interceptor being _too_ smart) could hunt our dreams in the future when we re-enable sending retrying.
Wouldn't this also be acceptable? when implementing retrying the message delivery we will not need to worry about the self deletion logic as this is already implemented?

**Comment by vitorhugods** [comment]
> 



> > Why delegate this to the Repository?
> > Also, I know we don't support this at the moment, but what about offline sending or scheduling of messages?
> > _i.e._ User tries to send while offline with timer `X`, then changes the timer to `Y`, and then tries to send another message.
> 
> What's the problem there :) ? The message will always use the most recent timer set on the conversation or user settings. Otherwise, as Mo pointed out, we might be hacking our own code/enforced settings.
> 
> > When retrying, the timer of both messages will be `Y`, violating the user's will because the timer is figured out on-the-fly.
> 
> I don't think this should be violating user's will, the message will be just following the current enforced settings with a little magic under the hood which is not evil per se imho.


This is the problem: `The message will always use the most recent timer set on the conversation or user settings` 😄. When the user explicitly tries to do otherwise.

> 
> > I would expect the UseCase to contain this logic and whatever the UseCase persists should be the law, and not to be altered by the sending mechanism.
> 
> I'm not necessarily against this approach, but following this will imply that we need to replicate this messageDao query for each use case we have now and any future ones that we add in the future (i.e. we decide at some point that reactions or pings should also be self deleted)

The duplication would be "getting the timer setting before persisting the message". 
So... two or three lines of code? Which we can have full control of.

> > I believe this kind of coupling and doubled responsibility (_i.e._ the interceptor being _too_ smart) could hunt our dreams in the future when we re-enable sending retrying.
> 
> Wouldn't this also be acceptable? when implementing retrying the message delivery we will not need to worry about the self deletion logic as this is already implemented?

It's a piece of business logic that we're delegating to the message sending mechanism instead of the one actually responsible for creating the message.

If I was asked where is this timer coming from, I'd check the UseCase first. As see it's not there, then have no clue where it comes from and have to dissect the sending mechanism to find a place that overwrites the original message.

----

As I said before, the UX design for this is debatable. 

_Maybe_ team - or group - enforced timer updates should overwrite pending messages composed while offline, _maybe_ not? The user hit the send button having one idea in mind.
But there's no benefit in not allowing the user to compose messages while offline with different timer settings. We've seen how sensitive users are when self-deleting messages don't work how they want recently.

By following what's persisted instead of making it on-the-fly, we have full flexibility on programming this behavior however we want.

When there's a group or team configuration change event, we can – if the UX design says so – update all the values in the DB for pending messages, before we come online and send them. Or the design might say to leave how it was, and we don't need to do anything.

But doing it on-the-fly brings no flexibility, and delegates the responsibility to an unexpected actor.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/Conversation.kt [hunk 1]
```diff
@@ -48,7 +48,8 @@ data class Conversation(
     val accessRole: List<AccessRole>,
     val creatorId: String?,
     val receiptMode: ReceiptMode,
-    val messageTimer: Duration? // enforced time by group conversation for self deleting messages, always NULL for 1on1 conversation
+    val messageTimer: Duration?, // enforced time by group conversation for self deleting messages, always NULL for 1on1 conversation
+    val userMessageTimer: Duration? // timer duration picked by the user in absence of team or conversation enforced timers
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/Conversation.kt:52]: Kdoc would be great here. So whoever presses F1 or Ctrl+Q on the IDE can see what's this about.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendingInterceptor.kt [hunk 2]
```diff
@@ -27,7 +27,7 @@ import com.wire.kalium.logic.functional.map
 import com.wire.kalium.logic.util.MessageContentEncoder
 
 interface MessageSendingInterceptor {
-    suspend fun prepareMessage(message: Message.Sendable): Either<CoreFailure, Message.Sendable>
+    suspend fun prepareMessage(originalMsg: Message.Sendable): Either<CoreFailure, Message.Sendable>
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendingInterceptor.kt:30]:
```suggestion
    suspend fun prepareMessage(originalMessage: Message.Sendable): Either<CoreFailure, Message.Sendable>
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendingInterceptor.kt [hunk 3]
```diff
@@ -36,6 +36,7 @@ class MessageSendingInterceptorImpl(
 ) : MessageSendingInterceptor {
 
     override suspend fun prepareMessage(message: Message.Sendable): Either<CoreFailure, Message.Sendable> {
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendingInterceptor.kt:38]:
```suggestion
    override suspend fun prepareMessage(originalMessage: Message.Sendable): Either<CoreFailure, Message.Sendable> {
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendingInterceptor.kt
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendingInterceptor.kt:1]: Or, alternatively, just revert these changes completely.

**Review by vitorhugods — APPROVED** [review]
A couple suggestions.

But LGTM in general.
